### PR TITLE
 RibonanzaNet-SS, folding util extraction

### PR DESCRIPTION
**Pull request by luxaritas**
## Summary
* Add support for the RibonanzaNet-SS folding engine
* Extract some generic folding utilities (pUnpaired, threshknot) into FoldUtil for easier reuse (and readability)

## Implementation Notes
* RibonanzaNet-SS inference setup was based on https://www.kaggle.com/code/shujun717/ribonanzanet-2d-structure-inference/
* We've introduced a new asset type, `onnx` (a standard format for ML models). Particularly notable is that our jest setup has to actually have imports of this asset provide its path so that it can be loaded to run in tests. I've changed all generic assets to do this (instead of just being an empty module) in case something else in the future requires the assets to be actually loaded.
* onnxruntime (responsible for doing the actual model inference) can only be run async, so we've had to ensure all codepaths which trigger folding operations can operate asynchronously (or, in the case of scripts, error when attempting to run synchronously). We use the existing PoseOp mechanism (originally built for allowing multistrand folding to yield to the main thread intermittently when not run from scripts) to do this, with some slight modifications.
* We have introduced new sometimes-sync-sometimes-async codepaths with function overloads, where a boolean determines whether it is marked async and returns a promise or is not marked async and returns the value directly. The actual implementation is always an async function, however the JS spec specifies that if no `await` is ever encountered, the function will run synchronously.
* When dot plot calculations are needed for constraints, instead of triggering the calculation in the constraint evaluate function (which has to run synchronously), we set a property on the constraint so that we can queue the calculation during the folding process (so that the data is already available when evaluate is run). We've also introduced the `skipMelt` parameter to `updateMeltingPointAndDotPlot` - this way if we're only running this update for the constraints, we don't have to run the extra fold/dot plot operations needed for the melt plot (which no constraints use)
* Unfortunately we can't make folding multithreaded, even though onnxruntime supports it (and we'd potentially gain something like a 2x speedup), as that requires a cross origin isolated browsing context, which we can't set up because of YouTube and Vimeo embeds :(

## Testing
* New unit tests
* Verified a couple samples were consistent with the kaggle notebook
* Misc interactive testing, including user feedback via the forums

### CODE REVIEW

**Comment by luxaritas** [comment]
TODO: change path to rnnet onnx file to engines-bin

File: src/eterna/mode/PoseEdit/PoseEditMode.ts [hunk 1]
```diff
@@ -2132,34 +2154,35 @@ export default class PoseEditMode extends GameMode {
...
         if (!specBox) return;
         this._specBox = specBox;
         this._specBox.setSpec(puzzleState);
-        this._specBox.closed.then(() => { this._specBox = null; });
+        await this._specBox.closed;
+        this._specBox = null;
     }
 
-    private updateSpecBox(): void {
+    private async updateSpecBox(): Promise<void> {
```
Comment (luxaritas) [inline src/eterna/mode/PoseEdit/PoseEditMode.ts:2169]: This seems ripe for a race condition

**Review by luxaritas — COMMENTED** [review]
Make sure to note in PR description:
* Async functions without await run synchronously
* Explain why we introduced skipMelt

**Review by tkaragianes — APPROVED** [review]
Reviewed in sync conversation with Jonathan, notes are in his review.

File: src/eterna/folding/RNNet.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        // to justify having everyone download it when not needed (which is excruciatingly
+        // slow on poor network connections eg in global regions)
+        if (this._session) return this._session;
+
+        const rnnetSs = await import('./engines/rnnet-ss.onnx');
+
+        ort.env.wasm.proxy = true;
+        ort.env.wasm.wasmPaths = {
+            'ort-wasm.wasm': new URL(
+                '../../../node_modules/onnxruntime-web/dist/ort-wasm.wasm',
```
[the remark this answers is not in the document]
Comment (luxaritas) [reply src/eterna/folding/RNNet.ts:172]: We do, because these files are not exported

**Comment by tkaragianes** [comment]
Closes #788.

**Pull request merged**